include/deal.II/base/timer_tree.h

```diff
@@ -0,0 +1,423 @@
+        print_name(pcout, offset, length);
+
+        // recursively print sub trees (decreasing the level and incrementing
+        // the offset)
+        for (auto it = sub_trees.begin(); it != sub_trees.end(); ++it)
+          {
+            (*it)->do_print_level(pcout,
+                                  level - 1,
+                                  offset + offset_per_level,
+                                  length);
+          }
+      }
+  }
+
+  void
+  print_name(ConditionalOStream const &pcout,
+             unsigned int const        offset,
+             unsigned int const        length) const
+  {
+    pcout << std::setw(offset) << "" << std::setw(length - offset) << std::left
+          << id;
+
+    pcout << std::endl;
+  }
+
+  void
+  print_own(ConditionalOStream const &pcout,
+            unsigned int const        offset,
+            unsigned int const        length,
+            bool const                relative = false,
+            double const              ref_time = -1.0) const
+  {
+    pcout << std::setw(offset) << "" << std::setw(length - offset) << std::left
+          << id;
+
+    Utilities::MPI::MinMaxAvg ref_time_data =
+      Utilities::MPI::min_max_avg(ref_time, MPI_COMM_WORLD);
+    double const ref_time_avg = ref_time_data.avg;
+
+    if (data.get())
+      {
+        double const time_avg = get_average_wall_time();
+
+        pcout << std::setprecision(precision) << std::scientific
+              << std::setw(10) << std::right << time_avg << " s";
+
+        if (relative)
+          pcout << std::setprecision(precision) << std::fixed << std::setw(10)
+                << std::right << time_avg / ref_time_avg * 100.0 << " %";
+      }
+
+    pcout << std::endl;
+  }
+
+  void
+  print_direct_children(ConditionalOStream const &pcout,
+                        unsigned int const        offset,
+                        unsigned int const        length,
+                        bool const                relative = false,
+                        double const              ref_time = -1.0) const
+  {
+    TimerTree other;
+    if (relative && sub_trees.size() > 0)
+      {
+        other.id = "Other";
+        other.data.reset(new Data());
+        other.data->wall_time = ref_time;
+      }
+
+    for (auto it = sub_trees.begin(); it != sub_trees.end(); ++it)
+      {
+        if ((*it)->data.get())
+          {
+            (*it)->print_own(pcout, offset, length, relative, ref_time);
+
+            if (relative)
+              other.data->wall_time -= (*it)->data->wall_time;
+          }
+      }
+
+    if (relative && sub_trees.size() > 0)
+      other.print_own(pcout, offset, length, relative, ref_time);
+  }
+
+  std::string id;
+
+  struct Data
+  {
+    Data()
+      : wall_time(0.0)
+    {}
+
+    double wall_time;
+  };
+
+  std::shared_ptr<Data> data;
+
+  std::vector<std::shared_ptr<TimerTree>> sub_trees;
+
+  static unsigned int const offset_per_level = 2;
+  static unsigned int const precision        = 2;
```

Comment:
Unless made `constexpr`, these `static` will also need to be defined in the source file.

Reply:
has been changed